src/utils/taint.js

```diff
@@ -0,0 +1,398 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { CompilerDiagnostic, FatalError } from "../errors.js";
+import type { Realm } from "../realm.js";
+import type { Descriptor, PropertyBinding, ObjectKind } from "../types.js";
+import {
+  DeclarativeEnvironmentRecord,
+  FunctionEnvironmentRecord,
+  ObjectEnvironmentRecord,
+  GlobalEnvironmentRecord,
+} from "../environment.js";
+import {
+  BoundFunctionValue,
+  ProxyValue,
+  SymbolValue,
+  AbstractValue,
+  EmptyValue,
+  FunctionValue,
+  PrimitiveValue,
+  Value,
+  ObjectValue,
+  NativeFunctionValue,
+} from "../values/index.js";
+import { IsDataDescriptor, TestIntegrityLevel } from "../methods/index.js";
+import type { BabelNodeExpression } from "babel-types";
+import invariant from "../invariant.js";
+
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
+  invariant(realm.generator, "a generator must exist while tracking a pure function");
+  let generator = realm.generator;
+  invariant(
```

Comment:
Why is that invariant true? Wouldn't it get violated when I simply pass `global` as an explicit argument to a abstract function? What's so special about `$GlobalObject`, what about other built-ins (thingies with an `.intrinsicName`)?

Reply:
If you pass global as an argument to an abstract function it won't be newly created so the visitor should have bailed out before you get here.

There isn't inherently special about global vs other intrinsics necessarily. However, if you do hit the global it'll pretty much reach the whole heap so it's better to bail out before it goes into some infinite loop.

We could check other intrinsics but since the intrinsic name is useful for other things (e.g. in the Node CLI there are a lot more intrinsics created than just built-in objects). It seems sketchy to assume something very specific that doesn't really have anything to do with the fact that it has an intrinsic name. E.g. I could call a function and have that object return an intrinsic object that can be recreated by calling it again.

The global object is one that is at least inherently to have been created before the additional function. There are more on that list but this is probably enough to help debugging.